### PR TITLE
Add liens to Terraform.

### PR DESCRIPTION
**Pull request by ndmckinley**
This resource is nonstandard in a number of ways.  This is a longish
read, settle in.  :)

First, it doesn't support `GET`s.  Instead, it supports a List operation.  This is
not unprecedented in MagicModules - we already have support for it in
Ruby, Puppet, and Ansible, and it's used for SQL Users, Flags and
Tiers, and DNS RecordSets.  This is what the `self_link_query` is for.

Second, it uses a different URL for `GET` and `DELETE`.  This *is*
unprecedented in MagicModules, and so I have done a hack - I used custom
code to rewrite the URL in `DELETE` before we send the request.  This
is not a good solution and I'm not happy with it.  Please suggest a better
one if you think it's worth it.

Third, it doesn't use `Operation`s at all.  This isn't unprecedented
in MagicModules - it just means we don't support timeouts and we
don't wait on operations.  But unfortunately we haven't written that
into Terraform, so I did that.

Fourth, it doesn't have a project.  We had some "no project" code in the
template preliminarily - we already didn't generate the "project" schema
entry if there was no project - but we still call `getProject` a lot and
that needed to be fixed.

Fifth, it doesn't have a self_link (because it doesn't support
`GET`).  We had some preliminary support for that, but it was broken
in the same way as the project support, and that needed to be fixed
as well.

Sixth, it has query parameters!  Those aren't unprecedented in magic
modules at all, you just handle them by making the self_link have the
necessary parameters, but currently terraform unconditionally concats
`?alt=json` to all our queries, and that's a bummer if there's already a
query parameter on there, so I fixed that.

Seventh, it is *very* finicky about how it accepts parameters - it
needs its project prefixed with `project/`, for instance, and it
only accepts project numbers instead of project names, and it stores
its name with the `liens/` prefix, which we generally prefer to
avoid.  I haven't fixed any of this at all - I think custom
expanders and flatteners are the right way, but I don't have a good
notion of how to build those reusably yet.  I'm still thinking on
that.  Resource works okay with that finickiness.

Eighth, resourceManager APIs are called all kinds of names.  It's called
`cloudresourcemanager` in the go generated code.  It's called
`gresourcemanager` in chef and puppet and ansible, which leads to
a generated name of `Resourcemanager` in Terraform, and that's wrong.
In general, it's messy.  So I added a shoddy hack to the top of
`resource.erb` to handle that.  Please suggest a better hack.

And finally, of course, I had to add in resourcemanager to be generated
by The Magician.

This is meant to address https://github.com/terraform-providers/terraform-provider-google/issues/1435.

-----------------------------------------------------------------
# [all]
Changes tangential to adding Liens resource (not yet supported here).
## [terraform]
Add liens resource.

Fixes #1435.

### CODE REVIEW

**Comment by ndmckinley** [comment]
@nelsonjr: I'd like your opinion on points number two and eight and `api.yaml`.
@danawillow and @rosbo: I'd like your opinion on everything else.  :)

File: products/resourcemanager/terraform.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      resource "google_resourcemanager_lien" "lien" {
+        parent = "projects/${google_project.project.number}"
+        restrictions = ["resourcemanager.projects.delete"]
+        origin = "machine-readable-explanation"
+        reason = "This project is very important to me!"
+      }
+      ```
+
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      post_create: templates/terraform/lien_post_create.erb
```
Comment (rosbo) [inline products/resourcemanager/terraform.yaml]: For my other PRs using custom code, I have been following a pattern where I put the files in a directory inside templates based on the custom_code injection point. 
https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/custom_flatten
https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/constants
https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/update_encoder

File: templates/terraform/modify_delete_url.erb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+url, err = replaceVars(d, config, "<%= object.__product.default_version.base_url -%>{{name}}")
```
Comment (rosbo) [inline templates/terraform/modify_delete_url.erb]: Consider adding a comment here with an explanation about why you added this block. I would take the explanation you added in this PR description. The generated code is a bit confusing to read the way it is.

Comment (rosbo) [reply templates/terraform/modify_delete_url.erb]: Another option is adding a `delete_url` and `create_url` to the Resource class.
If set `delete_url` and `create_url` are nil, default to the `base_url`.
Update the template to use these new urls.

File: templates/terraform/provider_gen.erb [hunk 3]
```diff
@@ -18,6 +18,11 @@ package google
 
 import "github.com/hashicorp/terraform/helper/schema"
 
+
+<% if product_ns == 'Resourcemanager'
+		product_ns = 'ResourceManager'
```
Comment (rosbo) [inline templates/terraform/provider_gen.erb]: Beautiful 🤦‍♂️

Comment (rosbo) [reply templates/terraform/provider_gen.erb]: In a different PR, we might evaluate adding a way to override a product name the same way we do it for properties or resources.

File: templates/terraform/resource.erb [hunk 4]
```diff
@@ -160,15 +181,60 @@ func resource<%= resource_name -%>Read(d *schema.ResourceData, meta interface{})
     return err
   }
 <% end -%>
+
+<% unless object.self_link_query.nil? -%>
```
Comment (rosbo) [inline templates/terraform/resource.erb]: I would move this to a separate template file and call `compile_template`

**Comment by ndmckinley** [comment]
FWIW: I think this is the best I can do for this resource right now.  Import, read, create, delete all work.  Some of it's a bit ugly and hacky but I think it's all correct.  There are two hack-fixes I can do:
* Make a 'name override' for products.
* Make `create_url` and `delete_url` and set them to `base_url` if nil.

Which, if either, do you think is worth doing in this PR?

**Comment by rosbo** [comment]
I would submit it as-is and make the change for product override and create/delete urls in separate PRs

File: products/resourcemanager/api.yaml [hunk 5]
```diff
@@ -93,3 +93,62 @@ objects:
             name: 'id'
             description: Id of the organization
 
+  - !ruby/object:Api::Resource
```
Comment (nelsonjr) [inline products/resourcemanager/api.yaml]: Suggestion: To make sure a new `api.yaml` works as expected we use to add to 2 providers at least. Even if they are separate PRs it is worth making sure nothing is inherently provider specific here.

File: templates/terraform/provider_gen.erb (hunk 3 above)
Comment (nelsonjr) [reply templates/terraform/provider_gen.erb]: You may want to extract the _meaning_ of the if into a concept and add it to `terraform.yaml`. Adding a product specific like this in the template is a major no-no.

File: templates/terraform/resource.erb [hunk 6]
```diff
@@ -19,11 +19,22 @@ package google
...
-  settable_properties = properties.reject(&:output)
-  api_name_lower = String.new(product_ns)
-  api_name_lower[0] = api_name_lower[0].downcase
+	# ResourceManager APIs need to be imported from api/cloudresourcemanager, need to
+	# be namespaced as ResourceManager, and need to be upstream as gresourcemanager,
+	# which leads to a product_ns of 'Resourcemanager'.  These are terraform-specific
+	# issues, so we just handle them here in the template.  If we start seeing this
+	# issue more than just in the ResourceManager APIs, we should invent a more general
+	# solution.
+	if product_ns == 'Resourcemanager'
```
Comment (nelsonjr) [inline templates/terraform/resource.erb]: ditto

**Comment by nelsonjr** [comment]
@ndmckinley regarding top PR list:

Two: That's okay when you have create and delete that are custom. For DNS we have a custom code for create that's different from delete.

Eight: One thing we could do... retire prefix in favor of a `short_name`. In the short name you have "Compute", "Resource Manager", etc. In P/C change the prefix method to build "g#{short_name.lower.no_spaces}". Which can be overridden later if needed. (although override does not affect non-resources [like product] today it should be easy to extend)

**Comment by ndmckinley** [comment]
@nelsonjr 
2)  Glad to hear it.  Going to stick with that, then, since it's the thing that's done elsewhere in MM, unless we end up seeing more of them than you'd expect.

8)  Good idea ... but I think this PR already fixes way too many things at once.  I'm getting uncomfortable continuing to add on.  I agree to make my next 2 tasks "remove the gross Resourcemanager -> ResourceManager hack" and "Add liens to puppet".

@rosbo @danawillow 
I fixed up number seven in the PR description (less reusably than I wanted, but it's the best I can do), and added on a number ten: You have to import this resource with its parent, and "parent" != "project", so I had to add the ability to use custom import IDs.

**Comment by ndmckinley** [comment]
Sure!

**Comment by nelsonjr** [comment]
That's reasonable. I suggest you file an issue so it does not get forgotten and TODO items in the code in the right places with the link to the issue.

File: products/resourcemanager/terraform.yaml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      pre_delete: templates/terraform/modify_delete_url.erb
+# This is for copying files over
+files: !ruby/object:Provider::Config::Files
```
[the remark this answers is not in the document]
Comment (ndmckinley) [reply products/resourcemanager/terraform.yaml]: Sure thing, done.

File: templates/terraform/decoders/avoid_meaningless_project_update.erb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  // Number, which is something that we can't address in a diffsuppress.
+  // Since we can't enforce a particular method of entering the project,
+  // we're just going to have to use whatever the user entered, whether
+  // it's project/projectName, project/12345, projectName, or 12345.
+  // The normal behavior of this method would be 'return res' - and that's
+  // what we'll fall back to if any of our conditions aren't met.  Those
+  // conditions are:
+  // 1) if the new or old values contain '/', the prefix of that is 'projects'.
+  // 2) if either is non-numeric, a project with that ID exists.
+  // 3) the project IDs represented by
```
[the remark this answers is not in the document]
Comment (ndmckinley) [reply templates/terraform/decoders/avoid_meaningless_project_update.erb]: Oh, hm, are we supposed to finish our sentences?  I'll keep that in mind for when I

**Comment by danawillow** [comment]
@nelsonjr @rosbo I think we should all agree where these custom code templates go- right now this PR has one in products/resourcemanager/helpers and others in templates/terraform/[post_import, pre_delete, etc.]

**Comment by rileykarson** [comment]
Hey @ndmckinley did we lose the tests here? They are in Terraform but not MM.

**Comment by ndmckinley** [comment]
Yep, that's right - didn't include the tests in autogeneration.  At the time we didn't know we were going to need to autogenerate all the tests eventually ... or at least I didn't.  :)

**Pull request merged**